packages/gatsby/src/commands/build.js

```diff
@@ -184,6 +215,35 @@ module.exports = async function build(program: BuildArgs) {
   }
   activity.done()
 
+  let deletedPageKeys = []
+  if (process.env.GATSBY_EXPERIMENTAL_PAGE_BUILD_ON_DATA_CHANGES) {
+    activity = report.activityTimer(`Delete previous page data`)
+    activity.start()
+    deletedPageKeys = pageDataUtil.collectRemovedPageData(
+      store.getState(),
+      cachedPageData
+    )
+
+    const removePages = deletedPageKeys.map(value => {
+      if (value === `/`) {
+        return fs.remove(`${program.directory}/public/index.html`)
+      } else {
+        return fs.remove(`${program.directory}/public${value}`)
```

Comment:
Ok, this is tricky.

Imagine you had pages `/blog`, `/blog/some-title` and you remove `/blog`

Then `public/blog` get's removed (with `/blog/some-title` with it). 

Ideally we would want to remove `index.html` files and if directory is empty remove directory as well (but that might be expensive when doing checks with `fs` so maybe there will be need to do some calculations in memory?

Don't have good answer here.

Reply:
Good point, 
I've updated it to just delete the file not the directory path.
We could check after remove if the directory now empty and then remove it. Would mean an extra action but clear public folder 🤔

Reply:
Yeah, we do need to delete directory if it's empty for sure, because hosting / server configuration might see that directory exists so it could show listing for empty directory in case of accessing `/blog` path.

I'm not sure how CDNs operate in those terms (do they track directories or just files?) and also if this is uniform for all CDNs or handling is CDN specific (and some might act one way and other in another)

Reply:
@pieh 
I've pushed up a solution, that from initial testing is working. 😄
If there are any subdirectories then the folder is not deleted using the callback.